lib/bulkhead/Bulkhead/Bulkhead.ts

```diff
@@ -0,0 +1,57 @@
+import { ExecutionQueue } from '../ExecutionQueue/ExecutionQueue';
+
+export class Bulkhead<T = any> {
+
+  private inExecution: number = 0;
+
+  constructor(
+    private readonly threshold: number,
+    private readonly executionQueue: ExecutionQueue<T>,
+  ) { }
+
+  public async run(method: (...args: any[]) => Promise<T>, args: any[]): Promise<T> {
+    if (this.inExecution < this.threshold) {
+      return this.execute(method, args);
+    }
+
+    return this.addToQueue(method, args);
+  }
+
+  private async execute(method: (...args: any[]) => Promise<T>, args: any[]) {
+    this.inExecution += 1;
+
+    const result = method(...args);
+
+    const afterExecution = async () => {
+      this.inExecution -= 1;
+      this.callNext();
+    };
+
+    result.then(afterExecution, afterExecution);
```

Comment:
What if the method never resolves? It will stay forever in the queue and potentially might block the queue. I think we might need an additional option to setup a timeout for the execution. Could we use the `@timeout` from current library? Either as a decorator, or directly as a method.